controllers/hvpa_util.go

```diff
@@ -0,0 +1,407 @@
+
+				if maxReplica < currentReplicas {
+					if (currentperReplicaCPU != 0 && currentperReplicaCPU < newPerReplicaCPU) && (currentperReplicaMem != 0 && currentperReplicaMem < newPerReplicaMem) {
+						// Paradoxical Scaling - block scaling for this container
+						log.V(2).Info("Paradoxical scaling: No scaling recommended because recommendation for horizontal scale is \"scale in\", while for vertical scale is \"scale up\"", "hvpa", hvpa.Namespace+"/"+hvpa.Name)
+						blockedScalingReason[container.Name] = hvpav1alpha1.BlockingReasonParadoxicalScaling
+						break
+					}
+				}
+				if maxReplica > finalReplica {
+					finalReplica = maxReplica
+				}
+				break
+			}
+		}
+	}
+	if finalReplica == 0 {
+		// scaling for all the containers is blocked
+		finalReplica = currentReplicas
+	}
+
+	newPodSpec = podSpec.DeepCopy()
+
+	for _, reco := range vpaStatus.Recommendation.ContainerRecommendations {
+		for i := range newPodSpec.Containers {
+			container := &newPodSpec.Containers[i]
+			if container.Name == reco.ContainerName {
+				currentperReplicaCPU := container.Resources.Requests.Cpu().MilliValue()
+				currentperReplicaMem := container.Resources.Requests.Memory().MilliValue()
+
+				newTotalCPUReco := int64(math.Max(float64(reco.Target.Cpu().MilliValue()*int64(currentReplicas)), float64(container.Resources.Requests.Cpu().MilliValue()*int64(desiredReplicas))))
+				newTotalMemReco := int64(math.Max(float64(reco.Target.Memory().MilliValue()*int64(currentReplicas)), float64(container.Resources.Requests.Memory().MilliValue()*int64(desiredReplicas))))
+
+				newCPUPerReplica := newTotalCPUReco / int64(finalReplica)
+				newMemPerReplica := newTotalMemReco / int64(finalReplica)
+
+				newCPUPerReplica, newMemPerReplica = adjustForBaseUsage(newCPUPerReplica, currentperReplicaCPU, newMemPerReplica, currentperReplicaMem, finalReplica, currentReplicas, hvpa)
+
+				if finalReplica > currentReplicas {
+					if currentperReplicaCPU > newCPUPerReplica && currentperReplicaMem > newMemPerReplica {
+						// Paradoxical Scaling
+						log.V(2).Info("Paradoxical scaling: current resources recommended because recommendation for horizontal scale is \"scale out\", while for vertical scale is \"scale down\"", "hvpa", hvpa.Namespace+"/"+hvpa.Name)
+						break
+					}
+					newCPUPerReplica = int64(math.Max(float64(currentperReplicaCPU), float64(newCPUPerReplica)))
+					newMemPerReplica = int64(math.Max(float64(currentperReplicaMem), float64(newMemPerReplica)))
+				}
+
+				newCPU := resource.NewScaledQuantity(newCPUPerReplica, resource.Milli)
+				_ = newCPU.String() // cache string q.s
+				newMem := resource.NewScaledQuantity(newMemPerReplica, resource.Milli)
+				newMem.SetScaled(newMem.ScaledValue(resource.Kilo), resource.Kilo)
+
+				target := corev1.ResourceList{
+					corev1.ResourceCPU:    *newCPU,
+					corev1.ResourceMemory: *newMem,
+				}
+				if reason, ok := blockedScalingReason[container.Name]; ok {
```

Comment:
Shouldn't blocking reason be revisited based on `finalReplica` and adjusted values? Especially, `BlockingReasonMinChange`. If one resource of one container has to be scaled vertically, we might as well apply all the changes blocked by `minChange` because disruption is happening anyway. WDYT?

Reply:
done in https://github.com/gardener/hvpa-controller/pull/76/commits/6bb252bf07f9b57484d991ccfa66a8615e177408